flaml/autogen/agent/user_proxy_agent.py

```diff
@@ -90,23 +96,109 @@ def _execute_code(self, code_blocks):
                 return exitcode, logs_all
         return exitcode, logs_all
 
-    def auto_reply(self, message, sender, default_reply=""):
+    @staticmethod
+    def _format_json_str(jstr):
+        """Remove newlines outside of quotes, and hanlde JSON escape sequences.
+
+        1. this function removes the newline in the query outside of quotes otherwise json.loads(s) will fail.
+            Ex 1:
+            "{\n"tool": "python",\n"query": "print('hello')\nprint('world')"\n}" -> "{"tool": "python","query": "print('hello')\nprint('world')"}"
+            Ex 2:
+            "{\n  \"location\": \"Boston, MA\"\n}" -> "{"location": "Boston, MA"}"
+
+        2. this function also handles JSON escape sequences inside quotes,
+            Ex 1:
+            '{"args": "a\na\na\ta"}' -> '{"args": "a\\na\\na\\ta"}'
+        """
+        result = []
+        inside_quotes = False
+        last_char = " "
+        for char in jstr:
+            if last_char != "\\" and char == '"':
+                inside_quotes = not inside_quotes
+            last_char = char
+            if not inside_quotes and char == "\n":
+                continue
+            if inside_quotes and char == "\n":
+                char = "\\n"
+            if inside_quotes and char == "\t":
+                char = "\\t"
+            result.append(char)
+        return "".join(result)
+
+    def _extract_args(self, input_string: str):
+        """Extract arguments from a json-like string and put it into a dict.
+
+        Args:
+            input_string: string to extract arguments from.
+
+        Returns:
+            a dictionary or None.
+        """
+        input_string = self._format_json_str(input_string)
+        try:
+            args = json.loads(input_string)
+            return args
+        except json.JSONDecodeError:
+            return {}
+
+    def _execute_function(self, func_call):
+        """Execute a function call and return the result.
+
+        Args:
+            func_call: a dictionary extracted from openai message at key "function_call" with keys "name" and "arguments".
+
+        Returns:
+            A tuple of (is_exec_success, result_dict).
+            is_exec_success (boolean): whether the execution is successful.
+            result_dict: a dictionary with keys "name", "role", and "content". Value of "role" is "function".
+        """
+        func_name = func_call.get("name", "")
+        func = self._function_map.get(func_name, None)
+
+        is_exec_success = False
+        if func is not None:
+            arguments = self._extract_args(func_call.get("arguments", ""))
+            if arguments is not None:
+                try:
+                    content = func(**arguments)
+                    is_exec_success = True
+                except Exception as e:
+                    content = f"Error: {e}"
+            else:
+                content = f"Error: Invalid arguments for function {func_name}."
```

Comment:
Is it better to send back the json decode error message?

Reply:
Yes, will revise and add tests.